### PR TITLE
Better error diagnostics for cyclic references

### PR DESCRIPTION
**Pull request by odersky**
We now suggest to compile with -explain-cyclic, in which case we give a trace of the forcings that led to the cycle.

The reason for the separate option is that maintaining a trace is not free so we should not be doing it by default.

### CODE REVIEW

**Comment by odersky** [comment]
I used this for #19398, which gave this:
```
-- [E046] Cyclic Error: tests/new/Test.scala:1:12 ------------------------------
1 |  def test: Option[Int] = Some(1)
  |            ^
  |Cyclic reference involving trait Seq
  |
  |The error occurred while trying to complete the info of method test
  |  which required to complete the info of class Option
  |  which required to read the definition of class Option in /Users/odersky/workspace/dotty/scala2-library-cc-tasty/../out/bootstrap/scala2-library-cc-tasty/scala-3.4.0-RC1-bin-SNAPSHOT-nonbootstrapped/scala2-library-cc-tasty_3-3.4.0-RC1-bin-SNAPSHOT.jar(scala/Option.tasty)
  |  which required to complete the info of type IterableOnce
  |  which required to read the definition of type IterableOnce in /Users/odersky/workspace/dotty/scala2-library-cc-tasty/../out/bootstrap/scala2-library-cc-tasty/scala-3.4.0-RC1-bin-SNAPSHOT-nonbootstrapped/scala2-library-cc-tasty_3-3.4.0-RC1-bin-SNAPSHOT.jar(scala/package.tasty)
  |  which required to complete the info of trait IterableOnce
  |  which required to read the definition of trait IterableOnce in /Users/odersky/workspace/dotty/scala2-library-cc-tasty/../out/bootstrap/scala2-library-cc-tasty/scala-3.4.0-RC1-bin-SNAPSHOT-nonbootstrapped/scala2-library-cc-tasty_3-3.4.0-RC1-bin-SNAPSHOT.jar(scala/collection/IterableOnce.tasty)
  |  which required to compute the base classes of trait Seq
  |  which required to compute the base classes of trait Iterable
  |  which required to complete the info of trait Iterable
  |  which required to read the definition of trait Iterable in /Users/odersky/workspace/dotty/scala2-library-cc-tasty/../out/bootstrap/scala2-library-cc-tasty/scala-3.4.0-RC1-bin-SNAPSHOT-nonbootstrapped/scala2-library-cc-tasty_3-3.4.0-RC1-bin-SNAPSHOT.jar(scala/collection/immutable/Iterable.tasty)
  |  which required to complete the info of trait Iterable
  |  which required to read the definition of trait Iterable in /Users/odersky/workspace/dotty/scala2-library-cc-tasty/../out/bootstrap/scala2-library-cc-tasty/scala-3.4.0-RC1-bin-SNAPSHOT-nonbootstrapped/scala2-library-cc-tasty_3-3.4.0-RC1-bin-SNAPSHOT.jar(scala/collection/Iterable.tasty)
  |  which required to compute the base classes of trait Seq
  |
  | Run with both -explain-cyclic and -Ydebug-cyclic to see full stack trace.
  |
  | longer explanation available when compiling with `-explain`
1 error found
```
After I turned the stacktrace on with -Ydebug-cyclic and correlated the two pieces of info, I found the root cause: It's because we now use more self types, which are annotated types, and which  create a link back from Iterable to Seq when loading Tasty. I'll try to tweak things on the cc side so that the self types are not needed.

File: compiler/src/dotty/tools/dotc/core/SymDenotations.scala [hunk 1]
```diff
@@ -167,12 +167,17 @@ object SymDenotations {
...
-          if (myFlags.is(Touched))
-            throw CyclicReference(this)(using ctx.withOwner(symbol))
-          myFlags |= Touched
-          atPhase(validFor.firstPhaseId)(completer.complete(this))
-        }
+        else
+          val traceCycles = CyclicReference.isTraced
+          try
+            if traceCycles then
+              CyclicReference.pushTrace("complete the info of ", symbol, "")
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/core/SymDenotations.scala:174]: `"complete the info of"` will be shown to users with `-explain-cyclic`. This terminology might not be clear to them. Maybe it should be something along the lines of `"computing the signature of"`

**Review by nicolasstucki — APPROVED** [review]
Otherwise, LGTM

File: compiler/src/dotty/tools/dotc/core/SymDenotations.scala (hunk 1 above)
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/SymDenotations.scala:174]:
```suggestion
              CyclicReference.pushTrace("compute the signature of ", symbol, "")
```

**Pull request merged**